File: src/cli/commands/diff-impact.ts
import { execFileSync } from "node:child_process";
import { existsSync } from "node:fs";
import type { Command } from "commander";
import { GraphQueries } from "../../core/graph/queries.js";
import { initDatabase } from "../../core/graph/schema.js";
import { Registry } from "../../core/registry.js";
import { log } from "../../utils/logger.js";
import { graphDbPath } from "../../utils/paths.js";

export function registerDiffImpactCommand(program: Command): void {
	program
		.command("diff-impact")
		.description("Show impact of current git changes")
		.option("--in <slug>", "Scope to a specific project")
		.option("--staged", "Only staged changes")
		.option("--depth <n>", "Max trace depth", "5")
		.action((opts: { in?: string; staged?: boolean; depth: string }) => {
			const registry = new Registry();
			const slug = opts.in ?? registry.findByPath(process.cwd())?.slug;

			if (!slug) {
				log.error("Not in a registered project.");
				process.exitCode = 1;
				return;
			}

			const project = registry.get(slug);
			if (!project) {
				log.error(`Project "${slug}" not found.`);
				process.exitCode = 1;
				return;
			}

			const dbPath = graphDbPath(slug);
			if (!existsSync(dbPath)) {
				log.error(`No graph for "${slug}". Run: graphmind build ${slug}`);
				process.exitCode = 1;
				return;
			}

			let changedFiles: string[];
			try {
				const args = opts.staged
					? ["diff", "--staged", "--name-only"]
					: ["diff", "--name-only", "HEAD"];
				const output = execFileSync("git", args, {
					cwd: project.path,
					encoding: "utf-8",
				});
				changedFiles = output
					.split("\n")
					.map((f) => f.trim())
					.filter(Boolean);
			} catch {
				log.error("Failed to run git diff. Are you in a git repository?");
				process.exitCode = 1;
				return;
			}

			if (changedFiles.length === 0) {
				log.dim("No changes detected.");
				return;
			}

			const db = initDatabase(dbPath);
			const q = new GraphQueries(db);
			const depth = Number.parseInt(opts.depth, 10);

			const allImpacted = new Set<string>();
			for (const file of changedFiles) {
				const impacted = q.impact(file, depth);
				for (const f of impacted) {
					allImpacted.add(f);
				}
			}

			db.close();

			console.log(`\n  Changed files (${changedFiles.length}):`);
			for (const f of changedFiles) {
				console.log(`    ● ${f}`);
			}

			if (allImpacted.size > 0) {
				console.log(`\n  Impacted files (${allImpacted.size}):`);
				for (const f of allImpacted) {
					if (!changedFiles.includes(f)) {
						console.log(`    ◦ ${f}`);
					}
				}
			} else {
				console.log("\n  No transitive impact detected.");
			}
			console.log();
		});
}

File: crates/graphmind-cli/src/commands/diff_impact.rs
use crate::paths;
use crate::resolve::resolve_project_slug;
use colored::Colorize;
use graphmind_db::queries::GraphQueries;
use graphmind_db::schema::init_database;
use std::process::Command;

pub fn diff_impact(slug: Option<&str>, staged: bool, depth: usize) {
    let slug = match resolve_project_slug(&[slug]) {
        Some(s) => s,
        None => {
            eprintln!(
                "{} No project specified and none could be resolved",
                "Error:".red().bold()
            );
            std::process::exit(1);
        }
    };

    let db_path = paths::graph_db_path(&slug);
    if !db_path.exists() {
        eprintln!(
            "{} No graph found for {}. Run 'graphmind build' first.",
            "Error:".red().bold(),
            slug
        );
        std::process::exit(1);
    }

    // Get changed files from git
    let changed_files = get_changed_files(staged);
    if changed_files.is_empty() {
        println!("{} No changed files detected", "!".yellow());
        return;
    }

    println!(
        "{} {} changed file(s):",
        ">>".cyan().bold(),
        changed_files.len().to_string().green()
    );
    for f in &changed_files {
        println!("  {}", f.yellow());
    }

    // Open the database and compute impact
    let db_path_str = db_path.to_string_lossy().to_string();
    let db = init_database(&db_path_str).unwrap_or_else(|e| {
        eprintln!("{} Failed to open database: {}", "Error:".red().bold(), e);
        std::process::exit(1);
    });

    let q = GraphQueries::new(&db);
    let mut all_impacted = std::collections::HashSet::new();

    for file in &changed_files {
        let impacted = q.impact(file, depth);
        for f in impacted {
            all_impacted.insert(f);
        }
    }

    // Remove changed files themselves from impact list
    for f in &changed_files {
        all_impacted.remove(f);
    }

    if all_impacted.is_empty() {
        println!("\n{} No additional files impacted", "OK".green().bold());
    } else {
        println!(
            "\n{} {} additional file(s) impacted (depth {}):",
            "!".yellow().bold(),
            all_impacted.len().to_string().red(),
            depth
        );
        for f in &all_impacted {
            println!("  {f}");
        }
    }
}

fn get_changed_files(staged: bool) -> Vec<String> {
    let args = if staged {
        vec!["diff", "--cached", "--name-only"]
    } else {
        vec!["diff", "--name-only", "HEAD"]
    };

    let output = Command::new("git")
        .args(&args)
        .output();

    match output {
        Ok(out) => {
            let stdout = String::from_utf8_lossy(&out.stdout);
            stdout
                .lines()
                .map(|l| l.trim().to_string())
                .filter(|l| !l.is_empty())
                .collect()
        }
        Err(_) => {
            // Fallback: try unstaged changes
            let fallback = Command::new("git")
                .args(["diff", "--name-only"])
                .output();
            match fallback {
                Ok(out) => {
                    let stdout = String::from_utf8_lossy(&out.stdout);
                    stdout
                        .lines()
                        .map(|l| l.trim().to_string())
                        .filter(|l| !l.is_empty())
                        .collect()
                }
                Err(_) => Vec::new(),
            }
        }
    }
}
